Java generation from CustomResourceDefinition yaml: field names containing dots, whitespace or quotes lead to compile error

### Describe the bug

If a CRD file contains a field with a name that contains a dot or a whitespace, the generated classes are not compiling. 
I found this [sanitize](https://github.com/fabric8io/kubernetes-client/blob/master/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java#L109-L133) method, which _seems_ to be responsible for cleaning the field names before creating the classes, but the cases dot/whitespace/quotes are not covered there.

### Fabric8 Kubernetes Client version

6.4.1

### Steps to reproduce

1. create a maven project with the following files:
    - pom.xml
        ```
          <?xml version="1.0" encoding="UTF-8"?>
        <project xmlns="http://maven.apache.org/POM/4.0.0"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
            <modelVersion>4.0.0</modelVersion>
        
            <groupId>org.example</groupId>
            <artifactId>property-dot-space</artifactId>
            <version>1.0-SNAPSHOT</version>
        
            <properties>
                <maven.compiler.source>11</maven.compiler.source>
                <maven.compiler.target>11</maven.compiler.target>
                <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
            </properties>
        
            <dependencies>
                <dependency>
                    <groupId>io.fabric8</groupId>
                    <artifactId>kubernetes-client</artifactId>
                    <version>6.4.1</version>
                </dependency>
                <dependency>
                    <groupId>io.fabric8</groupId>
                    <artifactId>generator-annotations</artifactId>
                    <version>6.4.1</version>
                </dependency>
            </dependencies>
        
            <build>
                <plugins>
        
                    <plugin>
                        <groupId>io.fabric8</groupId>
                        <artifactId>java-generator-maven-plugin</artifactId>
                        <version>6.4.1</version>
                        <executions>
                            <execution>
                                <goals>
                                    <goal>generate</goal>
                                </goals>
                                <id>crd-generator</id>
                            </execution>
                        </executions>
                        <configuration>
                            <source>${project.basedir}/src/main/resources</source>
                        </configuration>
                    </plugin>
                </plugins>
            </build>
        
        </project>
        ```
    - src/main/java/org/example/Main.java
        ```
        package org.example;
    
          public class Main {
              public static void main(String[] args) {
                  System.out.println("Hello world!");
              }
          }
        ```
    - src/main/resources/example-crd-dot.yaml
        ```
        apiVersion: apiextensions.k8s.io/v1
        kind: CustomResourceDefinition
        metadata:
          name: dotdummies.example.com
        spec:
          group: example.com
          versions:
            - name: v1
              served: true
              storage: true
              schema:
                openAPIV3Schema:
                  type: object
                  properties:
                    spec:
                      type: object
                      properties:
                        with.dot:
                          type: string
          scope: Namespaced
          names:
            plural: dotdummies
            singular: dotdummy
            kind: DotDummy
        ```
    - src/main/resources/example-crd-quote.yaml
        ```
        apiVersion: apiextensions.k8s.io/v1
        kind: CustomResourceDefinition
        metadata:
          name: quotedummies.example.com
        spec:
          group: example.com
          versions:
            - name: v1
              served: true
              storage: true
              schema:
                openAPIV3Schema:
                  type: object
                  properties:
                    spec:
                      type: object
                      properties:
                        with'quote:
                          type: string
          scope: Namespaced
          names:
            plural: quotedummies
            singular: quotedummy
            kind: QuoteDummy
        ```
    - src/main/resources/example-crd-space.yaml
        ```
        apiVersion: apiextensions.k8s.io/v1
        kind: CustomResourceDefinition
        metadata:
          name: spacedummies.example.com
        spec:
          group: example.com
          versions:
            - name: v1
              served: true
              storage: true
              schema:
                openAPIV3Schema:
                  type: object
                  properties:
                    spec:
                      type: object
                      properties:
                        with space:
                          type: string
          scope: Namespaced
          names:
            plural: spacedummies
            singular: spacedummy
            kind: SpaceDummy
        ```


2. run `mvn clean compile`


**Current behavior**
- build fails with the following output (paths anonymized):
    ```
    [INFO] Scanning for projects...
    [INFO] 
    [INFO] -------------------< org.example:property-dot-space >-------------------
    [INFO] Building property-dot-space 1.0-SNAPSHOT
    [INFO] --------------------------------[ jar ]---------------------------------
    [INFO] 
    [INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ property-dot-space ---
    [INFO] Deleting /exampleprojectpath/property-dot-space/target
    [INFO] 
    [INFO] --- java-generator-maven-plugin:6.4.1:generate (crd-generator) @ property-dot-space ---
    [INFO] 
    [INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ property-dot-space ---
    [INFO] Using 'UTF-8' encoding to copy filtered resources.
    [INFO] Copying 3 resources
    [INFO] 
    [INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ property-dot-space ---
    [INFO] Changes detected - recompiling the module!
    [INFO] Compiling 7 source files to /exampleprojectpath/property-dot-space/target/classes
    [INFO] -------------------------------------------------------------
    [ERROR] COMPILATION ERROR : 
    [INFO] -------------------------------------------------------------
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[11,24] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[11,30] <identifier> expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[13,26] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[13,27] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[14,20] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[14,21] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[17,25] '(' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[17,42] ',', ')', or '[' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,13] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,18] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,31] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,32] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[11,24] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[11,30] <identifier> expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[13,26] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[13,28] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[14,20] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[14,22] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[17,24] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[17,26] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[17,42] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,18] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,13] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,31] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,33] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[11,24] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[11,28] <identifier> expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[13,26] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[13,27] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[17,24] '(' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[17,25] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[17,44] wrong receiver parameter name
    [INFO] 32 errors 
    [INFO] -------------------------------------------------------------
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time:  1.123 s
    [INFO] Finished at: 2023-02-01T13:58:11+01:00
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.1:compile (default-compile) on project property-dot-space: Compilation failure: Compilation failure: 
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[11,24] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[11,30] <identifier> expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[13,26] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[13,27] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[14,20] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[14,21] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[17,25] '(' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[17,42] ',', ')', or '[' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,13] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,18] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,31] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/SpaceDummySpec.java:[18,32] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[11,24] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[11,30] <identifier> expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[13,26] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[13,28] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[14,20] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[14,22] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[17,24] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[17,26] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[17,42] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,18] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,13] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,31] unclosed character literal
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/QuoteDummySpec.java:[18,33] not a statement
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[11,24] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[11,28] <identifier> expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[13,26] ';' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[13,27] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[17,24] '(' expected
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[17,25] invalid method declaration; return type required
    [ERROR] /exampleprojectpath/property-dot-space/target/generated-sources/java/com/example/v1/DotDummySpec.java:[17,44] wrong receiver parameter name
    [ERROR] -> [Help 1]
    [ERROR] 
    [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
    [ERROR] Re-run Maven using the -X switch to enable full debug logging.
    [ERROR] 
    [ERROR] For more information about the errors and possible solutions, please read the following articles:
    [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
    ```
- generated `DotDummySpec.java` looks like:
    ```
    package com.example.v1;
    
    @com.fasterxml.jackson.annotation.JsonInclude(com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL)
    @com.fasterxml.jackson.annotation.JsonPropertyOrder({"with.dot"})
    @com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.fasterxml.jackson.databind.JsonDeserializer.None.class)
    @javax.annotation.processing.Generated("io.fabric8.java.generator.CRGeneratorRunner")
    public class DotDummySpec implements io.fabric8.kubernetes.api.model.KubernetesResource {
    
        @com.fasterxml.jackson.annotation.JsonProperty("with.dot")
        @com.fasterxml.jackson.annotation.JsonSetter(nulls = com.fasterxml.jackson.annotation.Nulls.SKIP)
        private String with.dot;
    
        public String getWith.dot() {
            return with.dot;
        }
    
        public void setWith.dot(String with.dot) {
            this.with.dot = with.dot;
        }
    }
    ```
- generated `QuoteDummySpec.java` looks like:
    ```
    package com.example.v1;
    
    @com.fasterxml.jackson.annotation.JsonInclude(com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL)
    @com.fasterxml.jackson.annotation.JsonPropertyOrder({"with'quote"})
    @com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.fasterxml.jackson.databind.JsonDeserializer.None.class)
    @javax.annotation.processing.Generated("io.fabric8.java.generator.CRGeneratorRunner")
    public class QuoteDummySpec implements io.fabric8.kubernetes.api.model.KubernetesResource {
    
        @com.fasterxml.jackson.annotation.JsonProperty("with'quote")
        @com.fasterxml.jackson.annotation.JsonSetter(nulls = com.fasterxml.jackson.annotation.Nulls.SKIP)
        private String with'quote;
    
        public String getWith'quote() {
            return with'quote;
        }
    
        public void setWith'quote(String with'quote) {
            this.with'quote = with'quote;
        }
    }
    ```
- generated `SpaceDummySpec.java` looks like:
    ```
    package com.example.v1;
    
    @com.fasterxml.jackson.annotation.JsonInclude(com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL)
    @com.fasterxml.jackson.annotation.JsonPropertyOrder({"with space"})
    @com.fasterxml.jackson.databind.annotation.JsonDeserialize(using = com.fasterxml.jackson.databind.JsonDeserializer.None.class)
    @javax.annotation.processing.Generated("io.fabric8.java.generator.CRGeneratorRunner")
    public class SpaceDummySpec implements io.fabric8.kubernetes.api.model.KubernetesResource {
    
        @com.fasterxml.jackson.annotation.JsonProperty("with space")
        @com.fasterxml.jackson.annotation.JsonSetter(nulls = com.fasterxml.jackson.annotation.Nulls.SKIP)
        private String with space;
    
        public String getWith space() {
            return with space;
        }
    
        public void setWith space(String with space) {
            this.with space = with space;
        }
    }
    
    ```
In the current state the generated classes are of no use and I do not see any valid workaround other than implementing these classes myself.

### Expected behavior

A successful maven build is expected which generated classes DotDummy and DotDummySpec, QuoteDummy and QuoteDummySpec, SpaceDummy and SpaceDummySpec and compiled them. 


### Runtime

other (please specify in additional context)

### Kubernetes API Server version

other (please specify in additional context)

### Environment

macOS

### Fabric8 Kubernetes Client Logs

_No response_

### Additional context

no kubernetes runtime or API server needed to reproduce this bug

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@
 * Fix #4798: fix leader election release on cancel
 * Fix #4815: (java-generator) create target download directory if it doesn't exist
 * Fix #4818: [java-generator] Escape `*/` in generated JavaDocs
+* Fix #4823: (java-generator) handle special characters in field names
 * Fix #4723: [java-generator] Fix a race in the use of JavaParser hitting large CRDs
 
 #### Improvements

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/AbstractJSONSchema2Pojo.java
@@ -131,7 +131,16 @@ public abstract class AbstractJSONSchema2Pojo {
       index = sanitized.indexOf('-');
     }
 
+    sanitized = sanitized.replace('.', '_');
+    sanitized = sanitized.replace(' ', '_');
+    sanitized = sanitized.replace('\'', '_');
+    sanitized = sanitized.replace('\"', '_');
+
     return sanitized;
+  }
+
+  public static String escapeQuotes(String str) {
+    return str.replace("\"", "\\\"").replace("\'", "\\\'");
   }
 
   public static AbstractJSONSchema2Pojo fromJsonSchema(

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JEnum.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JEnum.java
@@ -100,13 +100,15 @@ public class JEnum extends AbstractJSONSchema2Pojo {
       } catch (Exception e) {
         constantName = sanitizeEnumEntry(sanitizeString(k));
       }
+      String originalName = AbstractJSONSchema2Pojo.escapeQuotes(k);
+
       EnumConstantDeclaration decl = new EnumConstantDeclaration();
       decl.addAnnotation(
           new SingleMemberAnnotationExpr(
               new Name("com.fasterxml.jackson.annotation.JsonProperty"),
-              new StringLiteralExpr(k)));
+              new StringLiteralExpr(originalName)));
       decl.setName(constantName);
-      decl.addArgument(new StringLiteralExpr(k));
+      decl.addArgument(new StringLiteralExpr(originalName));
       en.addEntry(decl);
     }
 

--- a/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java
+++ b/java-generator/core/src/main/java/io/fabric8/java/generator/nodes/JObject.java
@@ -113,7 +113,8 @@ public class JObject extends AbstractJSONSchema2Pojo implements JObjectExtraAnno
   }
 
   private String getSortedFieldsAsParam(Set<String> list) {
-    List<String> sortedFields = list.stream().sorted().collect(Collectors.toList());
+    List<String> sortedFields = list.stream().map(AbstractJSONSchema2Pojo::escapeQuotes).sorted().collect(Collectors.toList());
+
     StringBuilder sb = new StringBuilder();
     sb.append("{");
     while (!sortedFields.isEmpty()) {
@@ -181,7 +182,7 @@ public class JObject extends AbstractJSONSchema2Pojo implements JObjectExtraAnno
       }
       buffer.addAll(gr.getTopLevelClasses());
 
-      String originalFieldName = k;
+      String originalFieldName = AbstractJSONSchema2Pojo.escapeQuotes(k);
       String fieldName = AbstractJSONSchema2Pojo.sanitizeString(k);
       String fieldType = prop.getType();
 

--- a/java-generator/core/src/test/java/io/fabric8/java/generator/CompilationTest.java
+++ b/java-generator/core/src/test/java/io/fabric8/java/generator/CompilationTest.java
@@ -57,6 +57,7 @@ class CompilationTest {
 
   private static Stream<Arguments> compilationTestData() {
     return Stream.of(
+        Arguments.of("dummy-crd.yml", 2),
         Arguments.of("crontab-crd.yml", 3),
         Arguments.of("keycloak-crd.yml", 50),
         Arguments.of("akka-microservices-crd.yml", 28),

--- a/java-generator/core/src/test/resources/dummy-crd.yml
+++ b/java-generator/core/src/test/resources/dummy-crd.yml
@@ -1,0 +1,48 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: dummies.example.com
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                zero:
+                  type: string
+                one.dot:
+                  type: string
+                two space:
+                  type: string
+                three'quote:
+                  type: string
+                four"doublequote:
+                  type: string
+  scope: Namespaced
+  names:
+    plural: dummies
+    singular: dummy
+    kind: Dummy

--- a/java-generator/it/src/it/escape-characters/invoker.properties
+++ b/java-generator/it/src/it/escape-characters/invoker.properties
@@ -1,0 +1,17 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+invoker.goals=test

--- a/java-generator/it/src/it/escape-characters/pom.xml
+++ b/java-generator/it/src/it/escape-characters/pom.xml
@@ -1,0 +1,88 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (C) 2015 Red Hat, Inc.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+            http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project>
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>escape-characters</artifactId>
+  <groupId>io.fabric8.it</groupId>
+  <version>0.0-SNAPSHOT</version>
+  <packaging>jar</packaging>
+
+  <properties>
+    <maven.compiler.source>@maven.compiler.source@</maven.compiler.source>
+    <maven.compiler.target>@maven.compiler.target@</maven.compiler.target>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>java-generator-integration-tests</artifactId>
+      <version>@project.version@</version>
+    </dependency>
+    <dependency>
+      <groupId>io.fabric8</groupId>
+      <artifactId>generator-annotations</artifactId>
+      <version>@project.version@</version>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-api</artifactId>
+      <version>@junit.version@</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.junit.jupiter</groupId>
+      <artifactId>junit-jupiter-engine</artifactId>
+      <version>@junit.version@</version>
+      <scope>test</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>io.fabric8</groupId>
+        <artifactId>java-generator-maven-plugin</artifactId>
+        <version>@project.version@</version>
+        <executions>
+          <execution>
+            <goals>
+              <goal>generate</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <source>src/test/resources/dummy-crd.yml</source>
+          <generatedAnnotations>false</generatedAnnotations>
+        </configuration>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-surefire-plugin</artifactId>
+        <version>3.0.0-M7</version>
+        <configuration>
+          <useFile>false</useFile>
+          <trimStackTrace>false</trimStackTrace>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+
+</project>

--- a/java-generator/it/src/it/escape-characters/src/test/java/io/fabric8/it/dummy/TestEscapeCharacters.java
+++ b/java-generator/it/src/it/escape-characters/src/test/java/io/fabric8/it/dummy/TestEscapeCharacters.java
@@ -1,0 +1,76 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.it.dummy;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.example.v1.Dummy;
+import com.example.v1.DummySpec;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.utils.Serialization;
+import org.junit.jupiter.api.Test;
+import io.fabric8.java.generator.testing.KubernetesResourceDiff;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.Files;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class TestEscapeCharacters {
+
+  @Test
+  void testDeserialization() {
+    // Arrange
+    Dummy sample =
+      Serialization.unmarshal(getClass().getResourceAsStream("/sample.yaml"), Dummy.class);
+
+    // Act
+    DummySpec spec = sample.getSpec();
+
+    // Assert
+    assertEquals("0", spec.getZero());
+    assertEquals("1", spec.getOne_dot());
+    assertEquals("2", spec.getTwo_space());
+    assertEquals("3", spec.getThree_quote());
+    assertEquals("4", spec.getFour_doublequote());
+  }
+
+  @Test
+  void testAgainstSample() throws Exception {
+    // Arrange
+    Path resPath = Paths.get(getClass().getResource("/sample.yaml").toURI());
+    String yamlContent = new String(Files.readAllBytes(resPath), "UTF8");
+    Dummy sample = new Dummy();
+    DummySpec spec = new DummySpec();
+    spec.setZero("0");
+    spec.setOne_dot("1");
+    spec.setTwo_space("2");
+    spec.setThree_quote("3");
+    spec.setFour_doublequote("4");
+    sample.setSpec(spec);
+    ObjectMeta om = new ObjectMeta();
+    om.setName("sample");
+    sample.setMetadata(om);
+    KubernetesResourceDiff diff = new KubernetesResourceDiff(yamlContent, Serialization.asYaml(sample));
+
+    // Act
+    List<JsonNode> aggregatedDiffs = diff.getListOfDiffs();
+
+    // Assert
+    assertEquals(0, aggregatedDiffs.size());
+  }
+}

--- a/java-generator/it/src/it/escape-characters/src/test/resources/dummy-crd.yml
+++ b/java-generator/it/src/it/escape-characters/src/test/resources/dummy-crd.yml
@@ -1,0 +1,48 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: dummies.example.com
+spec:
+  group: example.com
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                zero:
+                  type: string
+                one.dot:
+                  type: string
+                two space:
+                  type: string
+                three'quote:
+                  type: string
+                four"doublequote:
+                  type: string
+  scope: Namespaced
+  names:
+    plural: dummies
+    singular: dummy
+    kind: Dummy

--- a/java-generator/it/src/it/escape-characters/src/test/resources/sample.yaml
+++ b/java-generator/it/src/it/escape-characters/src/test/resources/sample.yaml
@@ -1,0 +1,26 @@
+#
+# Copyright (C) 2015 Red Hat, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+apiVersion: example.com/v1
+kind: Dummy
+metadata:
+  name: sample
+spec:
+  zero: "0"
+  one.dot: "1"
+  two space: "2"
+  three'quote: "3"
+  four"doublequote: "4"

